Prevent Setting SmtpClient.UseDefaultCredentials clearing Credentials

nabeelvalley opened the pull request:
Prevent setting `SmtpClient.UseDefaultCredentials = false` from clearing out `SmtpClient.Credentials` if set before that as noted in issue https://github.com/dotnet/runtime/issues/23779

nabeelvalley committed:
Ensure setting UseDefaultCredentials to False does not clear Credentials

nabeelvalley committed:
Merge branch 'smtp-credentials-fix'

nabeelvalley committed:
Remove reference comments

nabeelvalley committed:
Merge branch 'smtp-credentials-fix'

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/corefx?pullRequest=41271) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/corefx?pullRequest=41271) before we can accept your contribution.<br/><br/>:x: nabeelvalley [sign now](https://cla.dotnetfoundation.org/dotnet/corefx?pullRequest=41271)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/corefx?pullRequest=41271) it.</sub>

davidsh commented on src/System.Net.Mail/tests/Functional/SmtpClientTest.cs:
```diff
@@ -465,5 +465,95 @@ public void SendMail_DeliveryFormat_SubjectEncoded(bool useAsyncSend, bool useSe
         }
+
+        [Fact]
+        public void UseDefaultCredentialsUnset_ShouldReturn_CredentialsAsNull()
```
Some comments about naming tests:

```
<property or method under test>_<conditions>_<expected result>"
```

So, this test should named `Credentials_ctor_Null`.  This particular test is testing the value of the .Credentials property and not the .UseDefaultCredentials property.

davidsh replied:
So, please consider this pattern in the tests below and revise accordingly.

nabeelvalley replied:
Hi, thanks, I will do so

nabeelvalley committed:
Rename tests to align with convention

nabeelvalley committed:
Merge branch 'smtp-credentials-fix'

eiriktsarpalis commented on src/System.Net.Mail/src/System/Net/Mail/SmtpClient.cs:
```diff
@@ -238,15 +239,15 @@ public bool UseDefaultCredentials
                 }
 
-                _transport.Credentials = value ? CredentialCache.DefaultNetworkCredentials : null;
+                _useDefaultCredentials = value;
```
It looks like setting the `UseDefaultCredentials` property only impacts the credentials surfaced in SmtpClient but have no bearing on the underlying SmtpTransport state. Is this intentional?

nabeelvalley replied:
I didn't actually realize that, no it should probably set the transport state too since that is not being set anywhere, I'll take a look at that, how would you propose I add a test for that though?

nabeelvalley replied:
I'd just like to verify what behaviour we want here then:

1. If we set `Credentials`, we want the `_transport.Credentials` to be set to the`DefaultNetworkCredentials` if `UseDefaultCredentials == true`, otherwise we want it to be the newly set `Credentials` 
2. When setting `Credentials`, we only want the `_transpport.Credentials` to be set to the `Credentials` if we have `UseDefaultCredentials == false`

So with all that considered, when we use the `get` for `Credentials` do we want to return the Credentials we set or the current `_transport.Credentials`?

davidsh replied:
I think it would be best to first clarify the expected behavior from an API point of view rather than how that translates to code changes for the PR.

The SmptClient API has, unfortunately, 2 separate properties for specifying the credentials.  it has the 'UseDefaultCredentials' bool property and a separate 'Credentials' property.  In hindsight, it would have been better to have a single property.

That said, how should the properties interact with each other? That really is the topic of the original issue and what this PR is attempting to address.

The simple explanation for expected behavior that I think of is:

* Each property should be able to be 'get' and 'set' in any order and the last 'set' value should always be retrievable accurately regardless of the order in which these two properties are used.

* 'UseDefaultCredentials' has priority over the 'Credentials' property. If 'UseDefaultCredentials' is true, then default credentials are used regardless of what is set in the 'Credentials' property.  If 'UseDefaultCredentials' is false, then the 'Credentitals' property is used.

davidsh replied:
>how would you propose I add a test for that though?

Testing this with just unit tests is tricky. It really needs an end-to-end test. And that requires using an SMTP server that understand NTLM/Negotiate.  Negotiate is very hard because it requires the client and server to be joined on the same "domain" / "realm". We currently don't have a way to test that with our current CI (although work is in progress for expanding our Enterprise testing capabilities). NTLM is a little easier since the client doesn't have to be joined to a domain, but still requires some NTLM based SmtpServer.

We currently have a test SMTP server that is used in some of the unit tests. It's possible this could be expanded a little bit to at least test that the right NTLM challenge token was sent by SmtpClient.

But bottom line, we need some kind of "end to end" test here to validate that SmtpClient and related objects (like `_transport`) are correctly updated for the PR change.

nabeelvalley replied:
Okay great, so we're aligned on the expected behavior. 

I've been looking into testing this using a custom SMTP server but it does seem overkill to try to string one together to test that the credentials are being retained.

Can we not use reflection to access the `_transport.Credentials` data and compare that to what we'd set the `SmtpClient` with?

davidsh replied:
>Can we not use reflection to access the _transport.Credentials data and compare that to what we'd set the SmtpClient with?

We normally don't like to use reflection since it complicates the tests many times and relies on internal implementation details that could change in the future (thus breaking the test).

But given the low likelihood of this architecture changing and the difficulty is testing this, I'm ok with using reflection for creating a test to verify this change.

nabeelvalley committed:
Correctly set _transport.Credentials

nabeelvalley committed:
Include SmtpClient._transport.Credentials in tests, move tests to own file

nabeelvalley committed:
Merge branch 'master' of https://github.com/dotnet/corefx

nabeelvalley commented:
Hi @davidsh, I've added the `_transport` to be tested with the public credentials and everything seems to be working on my end, all the CI checks seem to be failing though

eiriktsarpalis commented:
> all the CI checks seem to be failing though

Seem to be unrelated errors affecting master as well. I would suggest waiting until CI issues have been resolved then building again.

davidsh commented:
/azp run

davidsh commented on src/System.Net.Mail/tests/Functional/CredentialTest.cs:
```diff
@@ -0,0 +1,152 @@
+        [Fact]
+        public void Credentials_UseDefaultCredentialsSetFalseBeforeCredentials_Credentials()
+        {
+            string userName = "user";
```
These variables for 'userName' and 'password' are repeated in several test methods. Also, using source code of the form:

```c#
```

will cause our static analysis tools to falsely flag this as a security problem (hard-coding passwords in code). Yes, it is a dummy password.

To avoid these problems, please:

1. create constants in the test class for these things to reduce duplicating code in the test methods
2. use something like `Guid` to set the dummy password

```c#
private readonly string UserName = "user";
private readonly string Password = Guid.NewGuid().ToString();
```

eiriktsarpalis commented on src/System.Net.Mail/tests/Functional/CredentialTest.cs:
```diff
@@ -0,0 +1,152 @@
+{
+    public class CredentialTest
+    {
+        private ICredentialsByHost GetTransportCredentials(SmtpClient client)
```
Nit: private implementation helpers should be at the bottom of the source file

eiriktsarpalis commented on src/System.Net.Mail/tests/Functional/CredentialTest.cs:
```diff
@@ -0,0 +1,152 @@
+
+namespace System.Net.Mail.Functional.Tests
+{
+    public class CredentialTest
```
If this is testing SmtpClient the test class should be named appropriately, e.g. `SmtpClientTests` or `SmtpClientCredentialsTests`. @davidsh can probably provide better guidance on naming conventions.

davidsh replied:
>@davidsh can probably provide better guidance on naming conventions.

I would name the class (and the filename) as `SmtpClientCredentialsTest`.

We usually always suffix the name with 'Test' not 'Tests' based on our following the guidance from the 'Art of Unit Testing" book (however, we have some deviations from this).

Since this test class and methods are testing both the 'UseDefaultCredentials' and 'Credentials' properties of the SmtpClient class, using "SmtpClientCredentialsTest" seems the most appropriate name.

nabeelvalley committed:
Change filename and refactor UN/PW

nabeelvalley commented:
Hi @davidsh @eiriktsarpalis, thanks for the feedback. I've made the suggested updates:

1. Use constants for username and password and a Guid for Password in Tests
2. Moved `GetTransportCredentials` helper function to end of file
3. Renamed class and file to `SmtpClientCredentialsTest`

davidsh commented:
@nabeelvalley Thanks for the PR!

danmosemsft commented:
@nabeelvalley if you are interested in more contributions, we have many issues marked 'up-for-grabs' 😄
